Bug 911 - Fix #scope{} and #group{} operator preplanning

rustyio opened the pull request:
This catches an issue found by Jon M. during a code review of the previous round of query planner changes.

Updates the #scope{} operator so that it rewrites the query to a #group{} operator during preplanning, rather than doing it later, allowing us to remove code in the #group{} operator that was accidentally duplicated.

rustyio committed:
Update #scope{} operator to correctly rewrite the query during preplanning. Remove duplicate code in #group{} operator.

jonmeredith commented:
This query (with default set to OR) adds and extra AND around the inner OR.  I think it should come out
OR["aaa","bbb","ccc"] but it is parsing as AND[OR["aaa","bbb","ccc"],NOT[OR[]]]]

(dev1@127.0.0.1)3> rp(search:explain("aaa bbb ccc")).  
{scope,#Ref<0.0.0.9669>,"search","value",
       {node,#Ref<0.0.0.9694>,'dev1@127.0.0.1',
             {intersection,#Ref<0.0.0.9670>,
                           [{node,#Ref<0.0.0.9692>,'dev1@127.0.0.1',
                                  {union,#Ref<0.0.0.9670>,
                                         [{term,#Ref<0.0.0.9673>,<<"aaa">>,
                                                [{'dev1@127.0.0.1',0},
                                                 {'dev1@127.0.0.1',0},
                                                 {'dev1@127.0.0.1',0}],
                                                0.0,1.0,#Fun<riak_search_op_term.default_filter.2>},
                                          {term,#Ref<0.0.0.9680>,<<"bbb">>,
                                                [{'dev1@127.0.0.1',0},
                                                 {'dev1@127.0.0.1',0},
                                                 {'dev1@127.0.0.1',0}],
                                                0.0,1.0,#Fun<riak_search_op_term.default_filter.2>},
                                          {term,#Ref<0.0.0.9687>,<<"ccc">>,
                                                [{'dev1@127.0.0.1',0},
                                                 {'dev1@127.0.0.1',0},
                                                 {'dev1@127.0.0.1',0}],
                                                0.0,1.0,#Fun<riak_search_op_term.default_filter.2>}]}},
                            {negation,#Ref<0.0.0.9670>,
                                      {node,#Ref<0.0.0.9693>,'dev1@127.0.0.1',
                                            {union,#Ref<0.0.0.9670>,[]}}}]}}}

rustyio committed:
Update to correctly handle prohibited terms in a group.

Fixes the following 2 cases:
1. Previously, when there were no prohibited terms, the query would have an empty operator.
2. Likewise, if there were *only* prohibited terms and no positive terms, the query would have an empty operator.

slfritchie commented:
Jon, is this merge'able?

FWIW, I used Jon's search:explain() example above with master as of changeset 7b04905170b0350abc51297124d7faf76c6cb528 (and not including patches from this pull request) and don't see the intersection/and that Jon's output shows ... but this may be pilot error? {shrug}

(riaksearch@127.0.0.1)1> rp(search:explain("aaa bbb ccc")). 
{scope,#Ref<0.0.0.2244>,"search","value",
       [{term,#Ref<0.0.0.2347>,<<"aaa">>,
              [{'riaksearch@127.0.0.1',0},
               {'riaksearch@127.0.0.1',0},
               {'riaksearch@127.0.0.1',0}],
              0.0,1.0,#Fun<riak_search_op_term.default_filter.2>},
        {term,#Ref<0.0.0.2354>,<<"bbb">>,
              [{'riaksearch@127.0.0.1',0},
               {'riaksearch@127.0.0.1',0},
               {'riaksearch@127.0.0.1',0}],
              0.0,1.0,#Fun<riak_search_op_term.default_filter.2>},
        {term,#Ref<0.0.0.2361>,<<"ccc">>,
              [{'riaksearch@127.0.0.1',0},
               {'riaksearch@127.0.0.1',0},
               {'riaksearch@127.0.0.1',0}],
              0.0,1.0,#Fun<riak_search_op_term.default_filter.2>}]}
ok

jonmeredith commented:
It looks like the original issue is fixed with the empty intersection / empty negation, I can still trigger a single entry union doing this query below.

(dev1@127.0.0.1)9>  rp(search:explain("(!aaa !bbb) !ccc")). 
{scope,#Ref<0.0.0.3494>,"search","value",
       {node,#Ref<0.0.0.3520>,'dev1@127.0.0.1',
             {intersection,#Ref<0.0.0.3495>,
                           [{node,#Ref<0.0.0.3511>,'dev1@127.0.0.1',
                                  {union,#Ref<0.0.0.3495>,
                                         [{negation,#Ref<0.0.0.3491>,
                                                    {node,#Ref<0.0.0.3510>,'dev1@127.0.0.1',
                                                          {union,#Ref<0.0.0.3491>,
                                                                 [{term,#Ref<0.0.0.3498>,<<"aaa">>,
                                                                        [{'dev1@127.0.0.1',0},
                                                                         {'dev1@127.0.0.1',0},
                                                                         {'dev1@127.0.0.1',0}],
                                                                        0.0,1.0,#Fun<riak_search_op_term.default_filter.2>},
                                                                  {term,#Ref<0.0.0.3505>,<<"bbb">>,
                                                                        [{'dev1@127.0.0.1',0},
                                                                         {'dev1@127.0.0.1',0},
                                                                         {'dev1@127.0.0.1',0}],
                                                                        0.0,1.0,#Fun<riak_search_op_term.default_filter.2>}]}}}]}},
                            {negation,#Ref<0.0.0.3495>,
                                      {node,#Ref<0.0.0.3519>,'dev1@127.0.0.1',
                                            {union,#Ref<0.0.0.3495>,
                                                   [{term,#Ref<0.0.0.3514>,<<"ccc">>,
                                                          [{'dev1@127.0.0.1',0},
                                                           {'dev1@127.0.0.1',0},
                                                           {'dev1@127.0.0.1',0}],
                                                          0.0,1.0,#Fun<riak_search_op_term.default_filter.2>}]}}}]}}}

rustyio committed:
Update #union and #intersection operators to unwrap when there is only a single child op.

rustyio commented:
Updated #union and #intersection operators to handle the case that Jon M. found above.

jonmeredith commented:
+1 confirmed fixed, did not try anything else.

beerriot commented:
+1 merge - explain examples cited above work correctly, and tests in tests/ directory pass
